data/features/featureList.yaml

```diff
@@ -0,0 +1,891 @@
+## Annotated examples
+## Just for documentation, not used by code. Please leave intact.
+examples:
+
+  # A feature gate that is alpha, enabled, and still configurable
+- name: ExampleFeature1
+  alpha:
+    fromVersion: v1.3
+    state: true # override the default state for this stage (false)
+  deprecated:
+    fromVersion: v1.18
+
+  # An experiment that didn't work out. After a while in alpha, it
+  # was deprecated and can no longer be enabled.
+- name: ExampleFeature2
+  alpha:
+    fromVersion: v1.15
+  deprecated:
+    fromVersion: v1.18
+
+  # A feature gate that is no longer available. This started as a
+  # beta featured but, unusually, one that was initially disabled
+  # by default.
+  # Later, it was enabled by default (still beta), eventually
+  # graduating to general availability.
+- name: ExampleFeature3
+  beta:
+    fromVersion: v1.4
+    state: false #
+    enabledFromVersion: v1.5
+  stable:
+    fromVersion: v1.7
+
+featureGateList:
+# Sorted alphabetically
+- name: Accelerators
+  alpha:
+    fromVersion: v1.6
+  deprecated:
+    fromVersion: v1.11
+- name: AdvancedAuditing
+  alpha:
+    fromVersion: v1.7
```

Comment:
👋 @sftim @tengqm 
I've been thinking on this a bit more and wonder if docs could run a trial (1.22 release) of generating the feature gates table (HTML) using a data file and shortcode. We'd have to provide some instructions about the process and expectations. If would be interesting to know if contributors prefer to edit and maintain a data file or the Markdown file. The Markdown file could remain in place in the content source. If this experiment does not work out, we can revert back to using Markdown (and update the content).

With this, what do you think about changing the rendered tables a bit?
Would it make sense to list the feature once in a single table with all of its states and versions. This would create one largish HTML table (use HTML not Markdown) and may mean that the table generation is a bit more complicated (columns, multiple rows).

To keep the shortcode source simple, make the data explicit (no calculations, all data appears in the data file) and the object definitions clear (no assumptions, though I think most contributors understand the difference between alpha and beta features). The shortcode bakes in the default rules (enabled or not). Another tool/template could use this data file to render the data differently.

Thinking about something like:

```yaml
# example
- name: Accelerators
  alpha:
       from: 1.6
       to: 1.9
       default: false # could skip, default is false
  beta:
       from: 1.10
       to:  1.20
       default: true # could skip, default is true
  ga:
       from: 1.2.1
```

Reply:
@kbhawkey As an engineer, I'd love to see all kinds of automation which help reduce manual labors and/or mistakes. However, this introduction of JSON data is not such a case because it is not buying us anything.

*It is not saving us manual labors*

The new JSON file is actually a different representation of the markdown table. The file has to be maintained and updated. We have to teach any contributor who is trying to update the content. Given that there are cases where the upstream developers intentionally break the protocol about "Alpha (disable)" -> "Beta (enable)" -> "GA/Deprecated", we will have to maintain a complicated logic to accommodate such deviations as exceptions. At the end of the day, the JSON data structure becomes verbose for no good. We will end up with a JSON data that is equivalent to the Markdown table. What we do accomplish is that we now have a template to maintain.

*It is not helping us reduce errors*

I do admit that we see some errors in PRs attempting to fix the markdown tables. Such mistakes are easy to spot and easy to fix. Translating the table into a JSON data file is not gonna save us any time on reviewing the content. It may look as if adding new changes would be simplified as just adding a new subsection. However, the devil details are still there. For example, shall/must we specify the "from" version? shall we remove the "to" version from the previous state when adding a new stage for a feature because that "to" version seems redundant, or we can ignore the "from" version in the new stage since that can be deduced? When a new gate is introduced, we are only supposed to add a "from" version, not the "to" version, is this always true? For new gate, we have to add a short description in a different file. We have to keep our eyes widely open for that ...

Reply:
@kbhawkey I don't mind a trial of generating feature gates using a shortcode.
In my view, the best time to cut over would be just after a release, before any PRs come in for the new _dev-1.23_ (or whatever branch). That way we can also sync the new approach into the dev branch with zero conflicts.

> what do you think about changing the rendered tables a bit?

I've deliberately kept the rendering identical / really close for **this** PR. If this PR merges, there's a lovely benefit: we can change how we present these data as much as we like, and that chang won't present a troublesome merge conflict for the release team.

We can make it into a single table (the shortcode already supports this) and we can if we want also decorate it with unobtrusive JavaScript / DHTML features like incremental search or filtering. All without disrupting the work on the dev branch for an upcoming relase.

> make the data explicit

There's clearly no single right answer - I'm a fan of my approach; you're recommending something different. Saying that the beta for `Accelerators` starts at 1.10 and also that the alpha ended in 1.9 feels almost tautological.
@kbhawkey it sounds like you're worried about encoding those rules in computer code. Can you elaborate on those concerns - are there risks that you have in mind?

Questions in my mind:
If we require the explicit data, should we later add a linter to encode those same default rules and object if there are violations? If we shouldn't do that linting, why not?


@tengqm suggested that:
> For new gate, we have to add a short description in a different file. We have to keep our eyes widely open for that ...

The closer we can get to having that YAML document listing all known feature gates, the easier it becomes to pick whether we reviewers _or some new automation_ are tasked to do that checking. It might become feasible to make sure that the rendered page or source Markdown at least includes every feature gate name known for a given upcoming release, so reviewers now only need to check that each description looks reasonable.

It is this kind of activity that I _do_ see as labor-saving and also helping improve the accuracy of the final page.

Now, we might have to live initially without all the automation we can eventually implement. I'm very happy to see this PR merge without that linting to check that each feature gate has a description, and add it later.

Reply:
> 
> 
> @kbhawkey As an engineer, I'd love to see all kinds of automation which help reduce manual labors and/or mistakes. However, this introduction of JSON data is not such a case because it is not buying us anything.
> 
> _It is not saving us manual labors_

Hi @tengqm @sftim .

I completely agree that pushing the feature gates table information into a data file does not automate the docs editing process.
The information is manually collected and added to the documentation. The information has to be manually reviewed.
During the closing period of a release, this page is frequently edited.
The greatest change is that contributors do not need to format the information and docs can possibly do more with an HTML table. Yes, a new template/shortcode adds maintenance work. If the shortcode is uncomplicated, this should be low overhead.

It would be less hassle to leave the tables as is and not try to improve how users work with this page.
I think the real reason for these changes is to set up the data and reuse in other ways (I may be wrong).
It has not been that difficult to help contributors edit this page, but it has been  a small bit of work during every release cycle.

I still think it might be worthwhile to evaluate how contributors use a data file and whether this file can generate the same or similar tables. The best time to evaluate this process is when the page is under greatest change (new features or contributors).
I think all the data has to be specified in the data file (no missing values or interpreted versions).
Yes, this is like duplicating the Markdown tables or creating a database.
The data set should mimic the current Markdown tables and the processing should not generate missing values hidden or patch up data (add, remove characters).
Possibly create a single table to avoid listing the feature gate name multiple times in the tables (first column spans multiple rows).

And there is the question about the description (in a different section on the page).
This is a long explanation without a full conclusion. The current page works, but it may be interesting to try out the data file. I think the current state of the data file needs to be simplified, though.